Assets/Labrys/FeatureLibrary.cs

```diff
@@ -0,0 +1,195 @@
+using Labrys.FeatureEditor;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+#if UNITY_EDITOR
+using UnityEditor;
+#endif
+using UnityEngine;
+
+namespace Labrys
+{
+    public class FeatureLibrary : ScriptableObject, IEnumerable<FeatureLibrary.Entry>, ISerializationCallbackReceiver, ICloneable
+    {
+        [field: SerializeField]
+        public string TargetDirectory { get; private set; }
+
+        private Dictionary<string, ushort> nameToFAID;
+        private FeatureAsset[] features;
+
+        public int Count => features.Length;
+
+        public FeatureAsset this[ushort faid] => Contains(faid) ? features[faid] : null;
+
+        public FeatureAsset this[string name]
+        {
+            get => this[GetId(name)];
+            private set => Add(name, value);
+        }
+
+#if UNITY_EDITOR
+        [MenuItem("Assets/Create/Labrys/Feature Library")]
+        private static void Create()
+        {
+            string saveDir = AssetDatabase.GetAssetPath(Selection.activeInstanceID);
+            if (!File.GetAttributes(saveDir).HasFlag(FileAttributes.Directory))
+                saveDir = saveDir.Replace(Path.GetFileName(saveDir), "");
+
+            string loadPath = $"Assets{EditorUtility.OpenFolderPanel("Select target Feature folder", saveDir, "").Replace(Application.dataPath, "")}";
+            string savePath = $"{saveDir}/NewFeatureLibrary.asset";
+
+            FeatureLibrary library = CreateInstance<FeatureLibrary>();
+            library.TargetDirectory = loadPath;
+            library.Refresh();
+
+            ProjectWindowUtil.CreateAsset(library, savePath);
+        }
+
+        public void Refresh()
+        {
+            string[] assetGUIDs = AssetDatabase.FindAssets($"t:{nameof(FeatureAsset)}", new string[] { TargetDirectory });
+            nameToFAID = new Dictionary<string, ushort>();
+            features = new FeatureAsset[assetGUIDs.Length];
+
+            foreach (string guid in assetGUIDs)
+            {
+                string path = AssetDatabase.GUIDToAssetPath(guid);
+                Add(path.Replace(TargetDirectory, "").Replace(".asset", "").TrimStart('/'),
+                    AssetDatabase.LoadAssetAtPath<FeatureAsset>(path));
+            }
+        }
+#endif
+        private FeatureLibrary()
+        {
+            nameToFAID = new Dictionary<string, ushort>();
+            features = new FeatureAsset[0];
+        }
+
+        private bool Add(string name, FeatureAsset feature)
+        {
+            if (name == null || feature == null)
+                return false;
+
+            features[nameToFAID.Count] = feature;
```

Comment:
`features` is an array, and it's only ever initialized to size 0. I don't see any lines where this array expands in size, so I think any time you call `Add`, you'll get an array index out of bounds exception.

Reply:
On further inspection, the only way this array gets created otherwise is by [de]serialization, so this is okay. This does limit the FeatureLibrary to being editor-only, and it can't be easily modified via code.